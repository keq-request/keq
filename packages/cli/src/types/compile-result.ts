export interface CompileResult {
  name: string
  path: string
  content: string
}
