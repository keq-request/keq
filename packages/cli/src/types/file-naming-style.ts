
export enum FileNamingStyle {
  camelCase = 'camelCase',
  capitalCase = 'capitalCase',
  constantCase = 'constantCase',
  dotCase = 'dotCase',
  headerCase = 'headerCase',
  noCase = 'noCase',
  paramCase = 'paramCase',
  pascalCase = 'pascalCase',
  pathCase = 'pathCase',
  sentenceCase = 'sentenceCase',
  snakeCase = 'snakeCase',
}
