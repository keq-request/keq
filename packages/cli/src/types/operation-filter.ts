/* eslint-disable @typescript-eslint/no-redeclare */
import { Type, type Static } from '@sinclair/typebox'


export const OperationFilter = Type.Object({
  // whether to generate files that do not exist
  append: Type.Optional(Type.Boolean({ default: false })),
  // whether to update files that already exist
  update: Type.Optional(Type.Boolean({ default: false })),

  // Only generate files of the specified operation method
  method: Type.Optional(Type.String()),
  // Only generate files of the specified operation pathname
  pathname: Type.Optional(Type.String()),
})


export type OperationFilter = Static<typeof OperationFilter>
