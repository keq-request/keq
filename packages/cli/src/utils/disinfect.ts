import { OpenAPI, OpenAPIV3 } from 'openapi-types'
import { removeUnnecessaryRef } from './remove-unnecessary-ref.js'
import { toSwagger3 } from './to-swagger3.js'
import { fixSwagger } from 'swagger-fix'
import { validateSwagger3 } from './validate-swagger3.js'

export async function disinfect(moduleName: string, swagger: OpenAPI.Document): Promise<OpenAPIV3.Document> {
  const swagger2 = fixSwagger(swagger)
  /**
   * NOTE: fixSwagger should before toSwagger3
   *       because toSwagger3 will replace Chinese with random numbers
   *       which will reduce readability
   */
  const swagger3 = await toSwagger3(swagger2)

  removeUnnecessaryRef(swagger3)

  return validateSwagger3(moduleName, swagger3)
}
