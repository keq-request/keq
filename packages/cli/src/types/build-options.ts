import { Type, type Static } from '@sinclair/typebox'
import { FileNamingStyle } from './file-naming-style'
import { OpenAPIV3 } from 'openapi-types'


export const BuildOptions = Type.Object({
  strict: Type.Optional(Type.Boolean({ default: false })),
  esm: Type.Optional(Type.Boolean({ default: false })),

  outdir: Type.String({ default: `${process.cwd()}/api` }),
  fileNamingStyle: Type.Enum(FileNamingStyle, { default: FileNamingStyle.snakeCase }),
  request: Type.Optional(Type.String()),


  modules: Type.Record(Type.String(), Type.Any()),
})

// eslint-disable-next-line @typescript-eslint/no-redeclare, no-redeclare
export interface BuildOptions extends Omit<Static<typeof BuildOptions>, 'operationId' | 'modules'> {
  modules: Record<string, OpenAPIV3.Document>
}
