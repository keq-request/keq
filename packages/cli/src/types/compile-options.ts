import { Type, type Static } from '@sinclair/typebox'
import { BuildOptions } from './build-options.js'
import { OpenAPIV3 } from 'openapi-types'


export const CompileOptions = Type.Intersect([
  Type.Omit(BuildOptions, ['modules']),

  Type.Object({
    moduleName: Type.String(),
    document: Type.Any(),
  }),
])

// eslint-disable-next-line @typescript-eslint/no-redeclare, no-redeclare
export interface CompileOptions extends Omit<Static<typeof CompileOptions>, 'document'> {
  document: OpenAPIV3.Document
}
