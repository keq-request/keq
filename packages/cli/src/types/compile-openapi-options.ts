import { CompileOptions } from './compile-options.js'
import { OpenAPIV3 } from 'openapi-types'


export interface CompileOpenapiOptions extends Omit<CompileOptions, 'filepath'> {
  document: OpenAPIV3.Document
}
