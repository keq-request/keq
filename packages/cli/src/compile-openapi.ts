import chalk from 'chalk'
import * as changeCase from 'change-case'
import * as fs from 'fs-extra'
import pMap from 'p-map'
import * as path from 'path'
import * as R from 'ramda'
import { getSafeOperationName } from './utils/get-safe-operation-name.js'
import jsonSchemaTemplateStr from './templates/json-schema/file.hbs'
import operationTemplateStr from './templates/openapi-core/operation.hbs'
import typeTemplateStr from './templates/openapi-core/type.hbs'

import Handlebars from 'handlebars'
import './handlebar/register-helper.js'
import './handlebar/register-partial.js'
import { CompileResult } from './types/compile-result.js'
import { CompileOpenapiOptions } from './types/compile-openapi-options.js'
import { FileNamingStyle } from './types/file-naming-style.js'
import { SupportedMethods } from './constants/supported-methods.js'


const readAndCompileTemplate = (tpl: string): ReturnType<typeof Handlebars.compile> => Handlebars.compile(tpl)
const templates = {
  t_schema: readAndCompileTemplate(jsonSchemaTemplateStr),
  t_operation: readAndCompileTemplate(operationTemplateStr),
  t_type: readAndCompileTemplate(typeTemplateStr),
}


function getModuleOutput(options: CompileOpenapiOptions): string {
  const moduleName = options.moduleName
  const fileNamingStyle: FileNamingStyle = options?.fileNamingStyle || FileNamingStyle.snakeCase
  const formatFilename = changeCase[fileNamingStyle]
  const outdir = options?.outdir || `${process.cwd()}/api`
  return path.join(outdir, formatFilename(moduleName))
}

export async function compile(options: CompileOpenapiOptions): Promise<CompileResult[]> {
  const esm = !!options.esm
  const moduleName = options.moduleName
  const document = options.document
  const fileNamingStyle: FileNamingStyle = options?.fileNamingStyle || FileNamingStyle.snakeCase
  const formatFilename = changeCase[fileNamingStyle]
  const output = getModuleOutput(options)
  const keq = options?.request ? path.relative(output, options.request) : 'keq'

  // const ignoredOperations: string[] = []
  const results: CompileResult[] = []
  if (document.components?.schemas && !R.isEmpty(document.components.schemas)) {
    for (const [name, jsonSchema] of R.toPairs(document.components.schemas)) {
      const context = {
        name,

        jsonSchema,
        document,

        esm,
        fileNamingStyle,
        keq,
      }

      const fileContent = templates.t_schema(context)
      let filename = formatFilename(name)
      filename = filename === 'index' ? 'index.schema' : filename
      const filepath = path.join(output, 'components', 'schemas', `${filename}.ts`)

      results.push({
        name: filename,
        path: filepath,
        content: fileContent,
      })
    }
  }

  if (document.paths && !R.isEmpty(document.paths)) {
    for (const [pathname, pathItem] of Object.entries(document.paths)) {
      if (!pathItem) continue

      for (const [m, operation] of Object.entries(pathItem)) {
        const method = m.toLowerCase()
        if (!SupportedMethods.includes(method)) {
          console.warn(chalk.yellow(`Method ${String(method).toUpperCase()} on path ${String(pathname)} cannot compiled, skipping`))
          continue
        }

        if (typeof operation === 'object' && !Array.isArray(operation)) {
          const context = {
            pathname,
            method,
            operation,

            document,
            moduleName,

            fileNamingStyle,
            esm,
            keq,
          }

          const filename = formatFilename(getSafeOperationName(pathname, method, operation))
          const operationFilepath = path.join(output, `${filename}.ts`)
          const operationTypeFilepath = path.join(output, 'types', `${filename}.ts`)

          const operationFileContent = templates.t_operation({ ...context })
          const operationTypeFileContent = templates.t_type({ ...context })


          // operation type file
          results.push({
            name: filename,
            path: operationTypeFilepath,
            content: operationTypeFileContent,
          })

          // operation file
          results.push({
            name: filename,
            path: operationFilepath,
            content: operationFileContent,
          })
        } else {
          console.warn(chalk.yellow(`Operation ${String(method)} on path ${String(pathname)} cannot compiled, skipping`))
        }
      }
    }
  }

  return results
}


async function genIndexFile(dir: string, options: CompileOpenapiOptions): Promise<void> {
  const names = await fs.readdir(dir)

  const pairs = await Promise.all(names.map(async (name) => [name, await fs.stat(path.join(dir, name))] as const))

  const filenames = pairs
    .filter(([name, fileStat]) => fileStat.isFile() && name.endsWith('.ts') && name !== 'index.ts')
    .map(([filename]) => filename)

  const content = filenames
    .map((filename) => `export * from "./${filename.slice(0, -3)}${options.esm ? '.js' : ''}"`)
    .join('\n')

  await fs.writeFile(path.join(dir, 'index.ts'), content)
}


export async function CompileOpenapi(option: CompileOpenapiOptions): Promise<void> {
  const results = await compile(option)

  await pMap(
    results,
    async (result) => {
      await fs.ensureFile(result.path)
      await fs.writeFile(result.path, result.content)
    },
    { concurrency: 10 },
  )

  const output = getModuleOutput(option)
  await genIndexFile(output, option)
  await genIndexFile(path.join(output, 'components', 'schemas'), option)
}
