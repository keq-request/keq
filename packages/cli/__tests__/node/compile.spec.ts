import { expect, test } from '@jest/globals'
import { OpenAPIV3 } from 'openapi-types'
import { compile } from '~/compile-openapi'
import * as swagger from '../swagger.json'
import { FileNamingStyle } from '~/types/file-naming-style'
import { disinfect } from '~/utils/disinfect'


test('compile swagger', async () => {
  const files = await compile({
    moduleName: 'test',
    document: await disinfect('test', swagger as unknown as OpenAPIV3.Document),
    outdir: './outdir',
    fileNamingStyle: FileNamingStyle.snakeCase,
    strict: false,
  })

  expect(files).toMatchSnapshot()
})

