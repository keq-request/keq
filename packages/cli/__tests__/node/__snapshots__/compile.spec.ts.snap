// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile swagger 1`] = `
[
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { ArrayOfStringsSchema } from "./array_of_strings_schema"
import type { ObjectWithSpecifiedTypeAdditionalPropertiesSchema } from "./object_with_specified_type_additional_properties_schema"
import type { ObjectWithAdditionalPropertiesSchema } from "./object_with_additional_properties_schema"


/**
 * @interface RootSchema
 * @export
 */
export interface RootSchema {
  "unknownProp"?: any
  "constProp": "constValue"
  "enumProp"?: "str" | 1 | {
    "key": true
  }
  /**
   * @deprecated
   */
  "deprecatedStringProp"?: string
  /**
   * Deprecated Number Prop
   *
   * @deprecated
   */
  "deprecatedNumberProp"?: string
  "emptyObjectProp"?: {
  }
  /**
   * This is String Prop
   */
  "stringProp"?: string
  /**
   * Multiple line description
   * This is Number Prop
   */
  "numberProp"?: number
  "booleanProp"?: boolean
  /**
   * @type int64
   */
  "integerProp"?: number
  "tupleProp"?: [string, number, boolean]
  "stringPropWithMaxLength"?: string
  "arrayOfStringsProp"?: (string)[]
  "arrayOfNullableRefProp"?: (ObjectWithSpecifiedTypeAdditionalPropertiesSchema | null)[]
  "refProp"?: ArrayOfStringsSchema
  "objectProp"?: {
    "stringPropRequired": string
    "stringPropNullable"?: string | null
    "refPropNullable"?: ObjectWithAdditionalPropertiesSchema | null
  }
}
",
    "name": "root_schema",
    "path": "outdir/test/components/schemas/root_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface ArrayOfStringsSchema
 * @export
 */
export type ArrayOfStringsSchema = (string)[]
",
    "name": "array_of_strings_schema",
    "path": "outdir/test/components/schemas/array_of_strings_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface ObjectWithAdditionalPropertiesSchema
 * @export
 */
export interface ObjectWithAdditionalPropertiesSchema {
  [key: string]: undefined | any
}
",
    "name": "object_with_additional_properties_schema",
    "path": "outdir/test/components/schemas/object_with_additional_properties_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface ObjectWithSpecifiedTypeAdditionalPropertiesSchema
 * @export
 */
export interface ObjectWithSpecifiedTypeAdditionalPropertiesSchema {
  /**
   * This is String Addition Prop
   */
  [key: string]: undefined | string
}
",
    "name": "object_with_specified_type_additional_properties_schema",
    "path": "outdir/test/components/schemas/object_with_specified_type_additional_properties_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface RefSchema
 * @export
 */
export type RefSchema = ArrayOfStringsSchema
",
    "name": "ref_schema",
    "path": "outdir/test/components/schemas/ref_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface NullableObjectSchema
 * @export
 */
export type NullableObjectSchema = {
  "stringProp"?: string
} | null
",
    "name": "nullable_object_schema",
    "path": "outdir/test/components/schemas/nullable_object_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface OneOfSchema
 * @export
 */
export type OneOfSchema = string | number | boolean
",
    "name": "one_of_schema",
    "path": "outdir/test/components/schemas/one_of_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface AnyOfSchema
 * @export
 */
export type AnyOfSchema = string | number | boolean
",
    "name": "any_of_schema",
    "path": "outdir/test/components/schemas/any_of_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

/**
 * @interface AllOfSchema
 * @export
 */
export type AllOfSchema = {
  "stringProp"?: string
} & {
  "numberProp"?: number
}
",
    "name": "all_of_schema",
    "path": "outdir/test/components/schemas/all_of_schema.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { KeqOperation } from 'keq'


export interface ResponseMap {
  "200": unknown
}


export type QueryParameters = {
  "requiredQueryProp": number
  "optionalQueryProp"?: number
}

export type RouteParameters = {
  "id": string
}

export type HeaderParameters = {
  "headerProp": string
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "get",
    "path": "outdir/test/types/get.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import { Keq } from "keq"
import { request } from "keq"
import type { RequestParameters, ResponseMap, Operation, QueryParameters, HeaderParameters, BodyParameters } from "./types/get"


export type GetRequestQuery = QueryParameters
export type GetRequestBody = BodyParameters
export type GetRequestHeaders = HeaderParameters


const pathname = "/:id"

export function get<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.get<ResponseMap[STATUS]>("/:id")
    .option('module', {
      name: "test",
      pathname,
    })

  const queryWrap = (value: any) => typeof value === 'boolean' ? String(value) : value

  if (arg && "requiredQueryProp" in arg) req.query("requiredQueryProp", queryWrap(arg["requiredQueryProp"]))
  if (arg && "optionalQueryProp" in arg) req.query("optionalQueryProp", queryWrap(arg["optionalQueryProp"]))
  if (arg && "headerProp" in arg && arg["headerProp"]) req.set("headerProp", arg["headerProp"])
  if (arg && "id" in arg) req.params("id", String(arg["id"]))

  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}

get.pathname = pathname
",
    "name": "get",
    "path": "outdir/test/get.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { KeqOperation } from 'keq'
import type { RootSchema } from "../components/schemas/root_schema"


export interface ResponseMap {
  "200": RootSchema
}


export type QueryParameters = {
}

export type RouteParameters = {
  "id": string
}

export type HeaderParameters = {
}

export type BodyParameters =(RootSchema)

export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "post_id",
    "path": "outdir/test/types/post_id.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import { Keq } from "keq"
import { request } from "keq"
import type { RequestParameters, ResponseMap, Operation, QueryParameters, HeaderParameters, BodyParameters } from "./types/post_id"


export type PostIdRequestQuery = QueryParameters
export type PostIdRequestBody = BodyParameters
export type PostIdRequestHeaders = HeaderParameters


const pathname = "/:id"

export function post__$id<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.post<ResponseMap[STATUS]>("/:id")
    .option('module', {
      name: "test",
      pathname,
    })

  if (arg && "id" in arg) req.params("id", String(arg["id"]))
  if (arg && "unknownProp" in arg) req.send({ "unknownProp": arg["unknownProp"] })
  if (arg && "constProp" in arg) req.send({ "constProp": arg["constProp"] })
  if (arg && "enumProp" in arg) req.send({ "enumProp": arg["enumProp"] })
  if (arg && "deprecatedStringProp" in arg) req.send({ "deprecatedStringProp": arg["deprecatedStringProp"] })
  if (arg && "deprecatedNumberProp" in arg) req.send({ "deprecatedNumberProp": arg["deprecatedNumberProp"] })
  if (arg && "emptyObjectProp" in arg) req.send({ "emptyObjectProp": arg["emptyObjectProp"] })
  if (arg && "stringProp" in arg) req.send({ "stringProp": arg["stringProp"] })
  if (arg && "numberProp" in arg) req.send({ "numberProp": arg["numberProp"] })
  if (arg && "booleanProp" in arg) req.send({ "booleanProp": arg["booleanProp"] })
  if (arg && "integerProp" in arg) req.send({ "integerProp": arg["integerProp"] })
  if (arg && "tupleProp" in arg) req.send({ "tupleProp": arg["tupleProp"] })
  if (arg && "stringPropWithMaxLength" in arg) req.send({ "stringPropWithMaxLength": arg["stringPropWithMaxLength"] })
  if (arg && "arrayOfStringsProp" in arg) req.send({ "arrayOfStringsProp": arg["arrayOfStringsProp"] })
  if (arg && "arrayOfNullableRefProp" in arg) req.send({ "arrayOfNullableRefProp": arg["arrayOfNullableRefProp"] })
  if (arg && "refProp" in arg) req.send({ "refProp": arg["refProp"] })
  if (arg && "objectProp" in arg) req.send({ "objectProp": arg["objectProp"] })

  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}

post__$id.pathname = pathname
",
    "name": "post_id",
    "path": "outdir/test/post_id.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { KeqOperation } from 'keq'


export interface ResponseMap {
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "get_simple",
    "path": "outdir/test/types/get_simple.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import { Keq } from "keq"
import { request } from "keq"
import type { RequestParameters, ResponseMap, Operation, QueryParameters, HeaderParameters, BodyParameters } from "./types/get_simple"


export type GetSimpleRequestQuery = QueryParameters
export type GetSimpleRequestBody = BodyParameters
export type GetSimpleRequestHeaders = HeaderParameters


const pathname = "/simple"

export function get__simple<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.get<ResponseMap[STATUS]>("/simple")
    .option('module', {
      name: "test",
      pathname,
    })


  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}

get__simple.pathname = pathname
",
    "name": "get_simple",
    "path": "outdir/test/get_simple.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { KeqOperation } from 'keq'


export interface ResponseMap {
  "200": {
}
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "post_simple",
    "path": "outdir/test/types/post_simple.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import { Keq } from "keq"
import { request } from "keq"
import type { RequestParameters, ResponseMap, Operation, QueryParameters, HeaderParameters, BodyParameters } from "./types/post_simple"


export type PostSimpleRequestQuery = QueryParameters
export type PostSimpleRequestBody = BodyParameters
export type PostSimpleRequestHeaders = HeaderParameters


const pathname = "/simple"

export function post__simple<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.post<ResponseMap[STATUS]>("/simple")
    .option('module', {
      name: "test",
      pathname,
    })


  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}

post__simple.pathname = pathname
",
    "name": "post_simple",
    "path": "outdir/test/post_simple.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { KeqOperation } from 'keq'


export interface ResponseMap {
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "delete_keyword",
    "path": "outdir/test/types/delete_keyword.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import { Keq } from "keq"
import { request } from "keq"
import type { RequestParameters, ResponseMap, Operation, QueryParameters, HeaderParameters, BodyParameters } from "./types/delete_keyword"


export type DeleteKeywordRequestQuery = QueryParameters
export type DeleteKeywordRequestBody = BodyParameters
export type DeleteKeywordRequestHeaders = HeaderParameters


const pathname = "/keyword"

export function deleteKeyword<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.delete<ResponseMap[STATUS]>("/keyword")
    .option('module', {
      name: "test",
      pathname,
    })


  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}

deleteKeyword.pathname = pathname
",
    "name": "delete_keyword",
    "path": "outdir/test/delete_keyword.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import type { KeqOperation } from 'keq'


export interface ResponseMap {
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters =({
  /**
   * file
   *
   * @type binary
   */
  "file": Blob | Buffer
})

export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "upload",
    "path": "outdir/test/types/upload.ts",
  },
  {
    "content": "// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.
// Code generated by keq-cli. DO NOT EDIT.

import { Keq } from "keq"
import { request } from "keq"
import type { RequestParameters, ResponseMap, Operation, QueryParameters, HeaderParameters, BodyParameters } from "./types/upload"


export type UploadRequestQuery = QueryParameters
export type UploadRequestBody = BodyParameters
export type UploadRequestHeaders = HeaderParameters


const pathname = "/upload"

/**
 * Upload File 
 * Description
 *
 * @summary Upload File
 */
export function upload<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.post<ResponseMap[STATUS]>("/upload")
    .option('module', {
      name: "test",
      pathname,
    })

  if (arg && "file" in arg && arg["file"]) req.attach("file", arg["file"])

  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}

upload.pathname = pathname
",
    "name": "upload",
    "path": "outdir/test/upload.ts",
  },
]
`;
